### ISSUE TITLE
VoiceOver is not working on macOS for Flutter websites and macOS apps

### ISSUE THREAD

**Issue by nilsreichardt** [flutter/flutter#115158]
## Description

Hey! 👋 I'm trying to get the macOS VoiceOver working for Flutter websites and Flutter macOS apps. However, it's not working. It's working for all other macOS apps or all other websites except that one created with Flutter.

https://user-images.githubusercontent.com/24459435/201398484-4aa5fc1b-02a8-4e08-86b5-482a783b035b.mp4

### Steps to reproduce

1. Go to https://gallery.flutter.dev
2. Enable VoiceOver in your macOS settings
3. Navigate with the tab key

<img width="780" alt="A screenshot of the VoiceOver setting" src="https://user-images.githubusercontent.com/24459435/201398676-30777e0c-cd26-4bbd-8d15-20abcd59c83a.png">

I would expect that VoiceOver is working like it shown in this video by Flutter: https://www.youtube.com/watch?v=A6Sx0lBP8PI

### Specs

<img width="698" alt="A screenshot of my 'About' mac settings windows" src="https://user-images.githubusercontent.com/24459435/201398935-3ddf7729-ec19-4278-8269-46461d97bb39.png">

**Comment by nilsreichardt** [comment]
https://user-images.githubusercontent.com/24459435/201410503-9c8fe320-5627-4aae-9fa5-4c040ba694a9.mov

I'm also unable to use the Ubunto (22.04) screen reader. It's working on the Firefox website (as shown in video) but on the Flutter gallery website.

**Comment by nilsreichardt** [comment]
Same for iOS on Flutter web. Twitter is working, but https://gallery.flutter.dev not.

https://user-images.githubusercontent.com/24459435/201469263-92d4192d-1fe3-4c6b-88c2-feabf6ad2d0c.MP4

**Comment by nilsreichardt** [comment]
Same for Android on Flutter web. Twitter is working, but https://gallery.flutter.dev/ not.


https://user-images.githubusercontent.com/24459435/201470096-404794aa-2361-4bba-a9a2-116a183c1cfb.mp4

**Comment by huycozy** [comment]
Hi @nilsreichardt, thanks for filing the issue. 
I checked this issue with [gallery](https://github.com/flutter/gallery) sample code, I see differences between the two platforms, the behavior is also different from widgets such as Buttons, TextFields:

- With macOS Flutter app: The VoiceOver doesn't work when hovering/clicking on Buttons, but works with pressing `Tab` key on TextFields. 
Also, instead of pressing `Tab` key, please try using this key combination: `Control + Option + Command + Up/Down`.

- With Web Flutter app (desktop): The VoiceOver works when using mouse (hovering) on Buttons and TextFields, but using `Tab` key doesn't work on Buttons.

<details>
<summary>Demo video (macos and web)</summary>

- <video src="https://user-images.githubusercontent.com/104349824/201610329-96475d66-3ded-48c3-9a4f-1621edb1e903.mp4"/>

- <video src="https://user-images.githubusercontent.com/104349824/201615448-6f7b4ed4-c3fe-40d2-8190-824a469ec764.mp4"/>

</details>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.3.8, on macOS 13.0 22A380 darwin-x64, locale en-VN)
    • Flutter version 3.3.8 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 52b3dc25f6 (23 hours ago), 2022-11-09 12:09:26 +0800
    • Engine revision 857bd6b74c
    • Dart version 2.18.4
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (3 available)
    • iPhone 14 Pro Max (mobile) • 332079AF-E895-4685-910E-7B1E18B0C6B8 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • macOS 13.0 22A380 darwin-x64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

```bash
[!] Flutter (Channel master, 3.6.0-3.0.pre.1, on macOS 13.0 22A380 darwin-x64, locale en-VN)
    • Flutter version 3.6.0-3.0.pre.1 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cb224d10e3 (26 minutes ago), 2022-11-13 21:57:38 -0500
    • Engine revision b7567b92c7
    • Dart version 2.19.0 (build 2.19.0-394.0.dev)
    • DevTools version 2.19.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.0 22A380 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```
</details>

Can you try again with mouse hovering and tab key actions to see if it's the same behavior? Maybe the issue is only caused by implementation in the Gallery source code or individual widgets.

**Comment by nilsreichardt** [comment]
Thanks for taking a look.

However, you can see in your recording (especially in web recording) that the screen reader does not read the content of the website (only when hovering with the mouse).

It should also read the things when using the TAB keys, like shown in this video by the Flutter team: https://youtu.be/A6Sx0lBP8PI. the author of the video says at the beginning that he uses the TAB key to navigate. However, using the TAB key does not result into the same behavior as shown in the video.

Also when we are putting the YouTube aside: How should a blind user navigate through the the Gallery app? The Gallery has accessibility support in the iOS and Android app. But when using the web, you can see the same behavior.

What you can also do: Compare this with non Flutter website / macOS apps. You will see that the TAB key is triggering the screen reader and a blind user would be able to use the app.

**Comment by huycozy** [comment]
@nilsreichardt Thanks for more information. Labeling this issue for further insights from the team.

**Comment by yjbanov** [comment]
Our expectation is that when VoiceOver is on the user would use the VoiceOver shortcuts to navigate the UI. By default there the semantics tree is disabled, but the user has the option to click on an accessibility placeholder to enable it. You can also enable it programmatically like this:

```dart
void main() {
  runApp(MyApp());
  RendererBinding.instance.setSemanticsEnabled(true);
}
```

However, it seems in some video recordings you are having trouble finding the placeholder. We should investigate, so I'll keep this issue open.

**Comment by nilsreichardt** [comment]
Thanks, @yjbanov. For everyone coming to this ticket in 2024 or later, you need to use the following code snippet:

```dart
void main() {
  runApp(MyApp());
  // Required on web/desktop to automatically enable accessibility features.
  WidgetsFlutterBinding.ensureInitialized().ensureSemantics();
}
```

The `RendererBinding` class doesn't exist anymore.

**Comment by orestesgaolin** [comment]
In my case the VoiceOver seems to work quite well on macOS once the focus is acquired by the window. In the video below you can see [app](https://github.com/orestesgaolin/flutter_desktop_sandbox) that doesn't use any Semantics widgets explicitly, only Focus and FocusableActionDetectors. I had few cases where VoiceOver would freak out completely but I think this was mostly due to screengrab software showing keystrokes on top.

https://github.com/user-attachments/assets/4c173164-e866-402f-ba87-d245a8a22531

**Comment by mdebbar** [comment]
VoiceOver has been working with Flutter apps for a while. If there are specific issues, please file them with details and reproduction steps.

For web, you have to enable semantics as suggested in [this comment](https://github.com/flutter/flutter/issues/115158#issuecomment-2094117364).
